### PR TITLE
feat: Rework authentication documentation to provide accurate information in terms of message workflows, permissions and auditability (PTL-1142)

### PR DESCRIPTION
**Pull request by jacinpoz**
Thank you for contributing to the documentation.

Do the changes you have made apply to both Current and Previous versions?
No, just current as there are slight differences between current and previous.

Have you checked all new or changed links?
Yes

Is there anything else you would like us to know?
No

For reference: 

  - We have an [internal contributions guide](https://www.notion.so/genesisglobal/Contributing-new-documentation-75953fb245f246ff872789035451a0c4)
  - We have a [style guide](https://www.notion.so/genesisglobal/Documentation-style-guide-5b04ec6fe12f4262b90d192effd8059b) 

**This week's exciting excerpts from the style guide**

- We write in UK English, with Oxford English Dictionary (OED) spellings, grammar and vocabulary.  

Use the present tense wherever possible. Only use another tense where it is strictly necessary.

- Present tense (preferred whenever possible): *The script creates a new folder.*
- Future tense (avoid whenever possible): *The script will create a new folder.*

### CODE REVIEW

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 1]
```diff
@@ -325,29 +325,63 @@ All requests below are capable of returning an error with a code of INTERNAL_ERR
 ---
 
 ## Pre-authentication
-Pre-authentication messages can be sent by a client without the user being logged in.
+Pre-authentication messages can be sent by a client without the user being logged in. There are four messages that can be used without being logged in.
 
-### Login preferences
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:330]: this flow never existed

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 2]
```diff
@@ -325,29 +325,63 @@ All requests below are capable of returning an error with a code of INTERNAL_ERR
...
-You must make sure that any connecting client knows the types of functionality that you have configured on the security module. For example, you could offer the client two ways of resetting user passwords: either via an administrator or by sending an email.  This choice can affect how the login dialog is displayed, so it is vital that the connecting client knows this before any user logs in.
-Currently, this is the only preference published.
-#### Request
-    MESSAGE_TYPE = EVENT_LOGIN_PREFS
-#### Response
-    MESSAGE_TYPE = EVENT_LOGIN_PREFS_ACK
-        DETAILS.PASSWORD_RESET_TYPE = ADMIN/EMAIL
+### Self-service password reset
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:330]: I am describing 3 different message flows that can happen before being authenticated here, as well as removing EVENT_LOGIN_PREFS given it doesn't exist.

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 3]
```diff
@@ -387,23 +424,47 @@ The error codes that can be returned are currently:
...
+
+```
+MESSAGE_TYPE = EVENT_LOGOUT
+DETAILS.USER_NAME = JohnWolf
+DETAILS.SESSION_ID = abb8f6be-8009-4370-a474-3a0ded4dc2cf
+```
 
 ## Post-authentication
-Once the user has been authenticated, the server expects heartbeat messages, as defined in the interval setting on the ACK message. If the GUI misses a configurable number of heartbeats, the session will automatically expire. In response to a heartbeat, the GUI will receive a list of available services and their details.
+Once the user has been authenticated, the server can optionally receive heartbeat messages, as defined in the interval setting on the ACK message. In response to a heartbeat, the GUI will receive a list of available services and their details, which could include multiple hostnames if the environment is configured to not use Consul as the cluster mode.
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:465]: This only applies if you are not building a website (websites can only target one host) and if the backend is not using consul as a cluster mode, so I had to clarify the explanation.

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 4]
```diff
@@ -427,20 +488,30 @@ These services should be contacted on the hosts in the order they are defined in
...
 In the Genesis low-code platform, there are profiles, users and rights. A profile is a group of users, which can be permissioned. For example, you could have a SALES_TRADER group in which all users must have the same permissions. In all cases where you specify either a right for a user/profile, or a user in a profile, the event represents what you want the entity to look like; i.e. if you amend a profile and don't supply a user that previously was part of that profile, then that user will be removed from that profile on the server.
 
 Note the following:
 
-* 2-phase validation is not currently supported
-* metadata is not supported on the following transactions.
-  User/profile `STATUS` field can be set to `ENABLED`, `DISABLED`, `PASSWORD_EXPIRED` and `PASSWORD_RESET`.
-  `PASSWORD_EXPIRED` should prompt the user to enter a new password.
-  `PASSWORD_RESET` should do the same but the server expects a blank "current password" field.
+* User `STATUS` field can be set to `ENABLED`, `DISABLED` or `PASSWORD_EXPIRED`. A User set up with `PASSWORD_EXPIRED` should prompt the user to enter a new password on next login.
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:513]: The information provided in the old docs was correct for many versions ago, but it is incorrect as of now. We do support 2-phase validation and we also provide metadata now.

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 5]
```diff
@@ -325,54 +325,116 @@ All requests below are capable of returning an error with a code of INTERNAL_ERR
...
     MESSAGE_TYPE = EVENT_LOGIN_AUTH
     DETAILS.USER_NAME = JohnWolf
-    DETAILS.PASSWORD = FullMoon1
+    DETAILS.PASSWORD = ******
 
 ### Login response
-If successful:
-
-    MESSAGE_TYPE = EVENT_LOGIN_AUTH_ACK
-    DETAILS.SYSTEM.DATE = 2014-06-18 12:27:01
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:353]: 2014... tells you where this example came from for sure and the reason why I had to really rework this page :)

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 6]
```diff
@@ -387,23 +449,47 @@ The error codes that can be returned are currently:
...
-- `INSUFFICIENT_CHARACTERISTICS` - can be provided if you have configured passwords to be successful if only 2 of 5.strength checks pass. Should be provided alongside "real" error codes.
+- `INSUFFICIENT_CHARACTERISTICS` - can be provided if you have configured passwords to be successful only if a predetermined set of strength checks pass. Should be provided alongside "real" error codes.
 - `ILLEGAL_SEQUENCE` - Numeric/alphabetic sequence detected.
 
-### Reset password
-This can only be called by an administrator; it simply specifies a user name and sets the password to blank.
+### Logout
 
-#### Reset request
-    MESSAGE_TYPE = EVENT_RESET_USER_PASSWORD
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:397]: reset user password is not in use anymore

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 7]
```diff
@@ -387,23 +449,47 @@ The error codes that can be returned are currently:
 - `INSUFFICIENT_CHARACTERS` - covers a few cases, so text field may be required, used for things like no digits provided when 1 digit is required.
 - `ILLEGAL_MATCH` - covers a few cases so text field may be required, used for things like repeating characters in password.
 - `ILLEGAL_WHITESPACE` - if password contains white space.
-- `INSUFFICIENT_CHARACTERISTICS` - can be provided if you have configured passwords to be successful if only 2 of 5.strength checks pass. Should be provided alongside "real" error codes.
+- `INSUFFICIENT_CHARACTERISTICS` - can be provided if you have configured passwords to be successful only if a predetermined set of strength checks pass. Should be provided alongside "real" error codes.
 - `ILLEGAL_SEQUENCE` - Numeric/alphabetic sequence detected.
 
-### Reset password
-This can only be called by an administrator; it simply specifies a user name and sets the password to blank.
+### Logout
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:455]: we never documented logout

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 8]
```diff
@@ -427,20 +513,76 @@ These services should be contacted on the hosts in the order they are defined in
...
+It is common for administrators to help users recover their account credentials by expiring the current User record `STATUS` with a new one-time password, and this action will force users to change it on their first login. In this case the message can optionally receive a one-time password like shown below:
+
+    MESSAGE_TYPE = EVENT_EXPIRE_USER_PASSWORD
+    DETAILS.USER_NAME = JohnWolf
+    DETAILS.PASSWORD = ******
+
+### Login details
+In the event the client needs to re-receive the information provided by the login response for whatever reason (i.e. re-reading user preferences), it is possible to send an EVENT_LOGIN_DETAILS message to the server. The response will be equivalent to the response received by the login message, without actively logging in the system for a second time.
+
+### Login details request
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:534]: we needed to document login details request

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 9]
```diff
@@ -427,20 +513,76 @@ These services should be contacted on the hosts in the order they are defined in
     DETAILS.SERVICE[1].HOST[1].NAME = genesisserv2
     DETAILS.SERVICE[1].HOST[1].PORT = 9002
 
+### Expire password
+A password can expire in three different scenarios: time based, user based or admin based. The end result is always the same, the user will need to change the password on the next login, as their previous one has now expired.
+
+#### Expire password request
+The password expiry mechanism can be triggered by sending a message like this:
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:520]: we needed to document expire password request

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 10]
```diff
@@ -427,20 +513,76 @@ These services should be contacted on the hosts in the order they are defined in
...
+    PERMISSION[10] = MFA_CONFIRM
+    PERMISSION[11] = MFA_CREATE
+    PERMISSION[12] = MFA_DISABLE
+    PERMISSION[13] = MFA_ENABLE
+    PROFILE[0] = ADMIN
+    PROFILE[1] = USER_ADMIN
 
 ### Rights polling
-The GUI can receive rights from a process called `AUTH_DATASERVER`. The view `USER_RIGHTS` displays all users and codes. A logged-in user should automatically set the Filter expression to be `USER_NAME`=='xxx' to receive push updates to user privileges.
+The GUI can receive rights from a process called `GENESIS_AUTH_DATASERVER`. The view `ALL_USER_RIGHTS` displays all users and codes. A logged-in user should automatically set the Filter expression to be `USER_NAME`=='xxx' to receive push updates to user privileges.
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:577]: wrong process and view names

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 11]
```diff
@@ -451,75 +593,34 @@ Note the following:
     DETAILS.NAME = SALES_TRADERS
     DETAILS.DESCRIPTION = Sales Traders
     DETAILS.STATUS = ENABLED
-    DETAILS.RIGHT[0].ID = 00000000000001RISP0
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:454]: IDs don't exist anymore and the names have changed.

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 12]
```diff
@@ -427,20 +513,76 @@ These services should be contacted on the hosts in the order they are defined in
...
 ## Entity management
 In the Genesis low-code platform, there are profiles, users and rights. A profile is a group of users, which can be permissioned. For example, you could have a SALES_TRADER group in which all users must have the same permissions. In all cases where you specify either a right for a user/profile, or a user in a profile, the event represents what you want the entity to look like; i.e. if you amend a profile and don't supply a user that previously was part of that profile, then that user will be removed from that profile on the server.
 
 Note the following:
 
-* 2-phase validation is not currently supported
-* metadata is not supported on the following transactions.
-  User/profile `STATUS` field can be set to `ENABLED`, `DISABLED`, `PASSWORD_EXPIRED` and `PASSWORD_RESET`.
-  `PASSWORD_EXPIRED` should prompt the user to enter a new password.
-  `PASSWORD_RESET` should do the same but the server expects a blank "current password" field.
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:443]: PASSWORD_RESET is deprecated

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 13]
```diff
@@ -451,75 +593,34 @@ Note the following:
...
 
 #### Amend response
 
 ```
-{
-  "GENERATED": [],
-  "MESSAGE_TYPE": "EVENT_ACK",
-  "SOURCE_REF": "1786d2ca-23fd-40c8-a52b-fe002a0fa1f6"
-}
+    MESSAGE_TYPE = EVENT_ACK
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:623]: all these events now just return EVENT_ACK (except some really special ones like LOGIN_AUTH or LOGIN_DETAILS)

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 14]
```diff
@@ -530,51 +631,78 @@ Note the following:
...
 #### Insert request
     MESSAGE_TYPE = EVENT_INSERT_USER
-    USER_NAME = jwolf
+    USER_NAME = JohnDoe
     DETAILS.USER_NAME = JohnWolf
     DETAILS.FIRST_NAME = John
     DETAILS.LAST_NAME = Wolf
     DETAILS.EMAIL_ADDRESS = john.wolf@genesis.global
     DETAILS.STATUS = ENABLED
-    DETAILS.RIGHT[0].ID = 00000000000001RISP0
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:545]: IDs don't exist and we don't allow rights to be inserted for a specific user, only profiles.

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 15]
```diff
@@ -530,51 +631,78 @@ Note the following:
...
-    DETAILS.USER_NAME = james
+    DETAILS.USER_NAME = JohnWolf
 
 ### Delete response
-    MESSAGE_TYPE = EVENT_DELETE_USER_ACK
+    MESSAGE_TYPE = EVENT_ACK
+
+## Full permissions list
+
+Most of the message workflows described in this page are permissioned based on a set of default rights which are provided in the form of a CSV file as part of the platform authentication distribution. Most permissions and their relationship to events have been explained in their own sections, and the whole table can be found below for reference:
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:680]: Adding the whole list of rights available for user management as well as special information about EXPIRE_PWD and CHANGE_PWD events.

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 16]
```diff
@@ -530,51 +631,78 @@ Note the following:
...
+
+A default Profile record is also provided named USER_ADMIN which contains all the previous rights.
+
+:::info
+Users can change their own password as well as expire their own password (assuming they are logged in). However, it is only possible to change/expire another user's password if you have administrator rights.
+:::
+
+## Event auditing
+
+Each one of these events is audited in one way or another, either by using the automatic mechanism provided at the table definition level (e.g. PROFILE, PROFILE_RIGHT, PROFILE_USER, PASSWORD_RESET, USER and USER_ATTRIBUTES), or by providing custom tables with the audit information.
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:703]: There was no information about event auditing, so I am adding it here.

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 17]
```diff
@@ -530,51 +631,84 @@ Note the following:
...
+
+In the second case scenario, we have automatic events to log changes in USER_AUDIT and USER_ATTRIBUTES when a password expires. And we also have specific handling for USER_LOGIN audits. The USER_LOGIN_AUDIT table will contain entries for the following events: LOGIN, LOGOUT, SESSION_EXPIRED, REJECTED (only available if a maximum number of user sessions has been configured), FAILED_LOGIN, and FAILED_LOGOUT (if an incorrect session ID or user name has been provided). 
+
+Additionally, all USER_LOGIN audit events are logged to a file at INFO level in the following format:
+
+    25 Jan 2024 16:31:18.823 12729 [dbCoroutinesContext-4 @coroutine#994] INFO  global.genesis.auth.manager.controller.LoginAuditController - AuditLoginEvent: UserLoginAudit{serialVersionUID='1',userLoginAuditId={not-set}, userName=JohnDoe, authAction=LOGIN, ipAddress=/127.0.0.1, reason=, recordId={not-set}, timestamp={not-set}}
```
Comment (jacinpoz) [inline docs/03_server/05_access-control/03_password_authentication.md:711]: Also adding a sample line of what a log statement would look like after a successful login.

File: docs/03_server/05_access-control/03_password_authentication.md (hunk 5 above)
Comment (wjhendry) [reply docs/03_server/05_access-control/03_password_authentication.md:353]: Hey, in my previous job, the system we were training people on started with (c)198-something

Comment (wjhendry) [reply docs/03_server/05_access-control/03_password_authentication.md:353]: but thanks for dragging my docs into the present day :)

File: docs/03_server/05_access-control/03_password_authentication.md (hunk 3 above)
Comment (wjhendry) [reply docs/03_server/05_access-control/03_password_authentication.md]: I have now changed the heading and moved this to the end of the post-auth section, to ensure that it is not a BAU consideration

File: docs/03_server/05_access-control/03_password_authentication.md [hunk 18]
```diff
@@ -451,75 +593,34 @@ Note the following:
...
 ### Amend profile
 
 #### Amend request
-  In the example below, the logged-in user (in the second line) is JohnWolf, who is modifying the profile of JaneDoe to give her the profile name JANE SMITH. 
+
+This amend request removes the "ORDAM" right and the "james" user name from the SALES_TRADERS profile.
 
 ```
-    {
-  "SOURCE_REF": "1786d2ca-23fd-40c8-a52b-fe002a0fa1f6",
```
[the remark this answers is not in the document]
Comment (wjhendry) [reply docs/03_server/05_access-control/03_password_authentication.md:473]: Thanks, @jacinpoz  . Please check lines 619-622

File: docs/03_server/05_access-control/03_password_authentication.md (hunk 3 above)
Comment (wjhendry) [reply docs/03_server/05_access-control/03_password_authentication.md]: @jacinpoz please check lines 563-588

**Review by wjhendry — APPROVED** [review]

**Pull request merged**